Add a module supporting JSON output from xml node.

kyoshinaga opened the pull request:
The JSONUtil.scala is added into the util directory. 
And some switching step is added into pipeline.scala.

Commit:
Just add JSONUtil object

Commit:
[add/update] JSONUtil object.

. Add new option, `-json` to pipeline.Pipeline.
. Modyfy following methods.
 . Pipeline.run
 . Pipeline.runFromStdin
 . Pipeline.shell

Commit:
[add] New option, "-outputFormat" to Pipeline

. Add new option value, "-outputFormat".
. Add message.
. Modify following methods.
 . Pipeline.run
 . Pipeline.runFromStdin
 . Pipeline.shell

hiroshinoji commented on src/main/scala/jigg/pipeline/Pipeline.scala:
```diff
@@ -32,10 +32,11 @@ class Pipeline(val properties: Properties = new Properties) extends PropsHolder
   @Prop(gloss="Property file") var props = ""
   @Prop(gloss="Input file; if omitted, read from stdin") var file = ""
-  @Prop(gloss="Output file; if omitted, `file`.xml is used. Gzipped if suffix is .gz") var output = ""
+  @Prop(gloss="Output file; if omitted, `file`.xml is used. Gzipped if suffix is .gz. if JSON mode is selected, suffix is .json") var output = ""
```
Typo: if JSON => If JSON

hiroshinoji commented on src/main/scala/jigg/pipeline/Pipeline.scala:
```diff
@@ -202,22 +203,38 @@ Currently the annotators listed below are installed. See the detail of each anno
-    val printer = new scala.xml.PrettyPrinter(500, 2)
+    // If option of json is true, the output of this method is formatted JSON.
+    prop("outputFormat") match {
+      case Some(x) if x == "json" => {
```
The following expression is more concise:

``` scala
case Some("json") => 
```

hiroshinoji commented on src/main/scala/jigg/pipeline/Pipeline.scala:
```diff
@@ -202,22 +203,38 @@ Currently the annotators listed below are installed. See the detail of each anno
-    // XML.save(in + ".xml", xml, "UTF-8")
-    val printer = new scala.xml.PrettyPrinter(500, 2)
+    // If option of json is true, the output of this method is formatted JSON.
+    prop("outputFormat") match {
```
Due to the mechanism of `@Prop` (sorry this is not yet well documented), `outputFormat` variable is automatically overwritten with the value of `prop('outputFormat')` (the default value is `xml`). Thus, the pattern matching with Some is unnecessary:

``` scala
outputFormat match {
  case "json" => ...
  case _ => ...
}
```

hiroshinoji commented on src/main/scala/jigg/pipeline/Pipeline.scala:
```diff
@@ -237,7 +254,12 @@ Currently the annotators listed below are installed. See the detail of each anno
           case _ => IOUtil.openOut(output)
         }
-        writeTo(writer, xml)
+        prop("outputFormat") match {
```
The same comment as above.

hiroshinoji commented on src/main/scala/jigg/util/JSONUtil.scala:
```diff
@@ -0,0 +1,133 @@
+  object XMLParser {
+    // If text node has newline character, Node.child method doesn't work collectly.
+    def trimNode(x: Node): Node = 
+      XML.loadString(x.toString.split("\n").mkString)
```
I don't think this is the correct thing to do. In the command below:

``` sh
$ echo "dogs\nrun" | java -cp target/jigg-assembly-0.6.jar:jar/stanford-corenlp-3.6.0-models.jar jigg.pipeline.Pipeline -annotators "corenlp[tokenize,ssplit]" -outputFormat json
```

The result is:

``` json
{"root":{ "_":[
...
{ "sentence":{
"text":"dogsrun",
...
{ "token":{
"form":"dogs",
"id":"t0",
"characterOffsetBegin":"0",
"characterOffsetEnd":"4"}},
{ "token":{
"form":"run",
"id":"t1",
"characterOffsetBegin":"5",
"characterOffsetEnd":"8"}}]
...
```

and offsetBegin = 5 for "run" is wrong (we cannot recover the original text from text).

What is the problem when newline character is included in text? Perhaps the problem can be solved with `XMLUtil.text` method? This tries to extract just the original text from the given node.

hiroshinoji commented:
LGTM except some minor comments, listed above in the code. One problem is handling of newline characters in text, which should be resolved in some way...

kyoshinaga commented:
Sorry for my mistakes and thanks a lot for your comments.
I'll revise my code and consider other ways to resolve the problem about the newline characters.

kyoshinaga commented:
> Perhaps the problem can be solved with XMLUtil.text method?

Sorry, I couldn't find the method before making my code.
I think it can solve this problem.

Commit:
[update] pipeline/Pipeline.scala

- Correct typo.
- Modify pattern matchings.

hiroshinoji commented:
The method can be found here:
https://github.com/mynlp/jigg/blob/develop/src/main/scala/jigg/util/XMLUtil.scala#L82

Commit:
[update] util/JSONUtil.scala

- Delete XMLParser.trimNode
- Delete XMLParser.getText
- Add XMLUtil.text to JSONUtil.serializing

kyoshinaga commented:
Thank you so much for your comment. The problem is solved by using the method.
And I revised my code in accordance with your comments.

Commit:
[update] util/JSONUtil.scala

- Modify the output format of JSON.
- Added xml.Utility.escape into JSONUtil.serializing to escape the
  special characters.

kyoshinaga commented:
I modified the output format of json as below.
- A name of tags, list of child elements, and attributions are stored into a single dictionary. 
- The name of tag is stored as value of '.tag' key.
- The list of child elements stored as value of  '.child' key.
  - The '.child' includes sub node of 

For example, Input as below:

``` xml
<sentence id="s1">
  <aaa/>
  <bbb/>
</sentence> 
```

The result is:

``` json
{ ".tag": "sentence",
  "id": "s1",
  ".child": [{ ".tag": "aaa" },
             { ".tag": "bbb" }]
}
```

In addition, xml.Utility.escape method added into JSONUtil.scala to escape the special characters.
Then I pushed above modification to my repository.
I'm sorry to bother you, but could you review it again?
Or, should I send PR to you again?

hiroshinoji commented:
Thank you very much! New output format looks much better. I will merge it.
